blockchain/blockchain.go

```diff
@@ -645,25 +645,28 @@ func (bc *BlockChain) setHeadBeyondRoot(head uint64, root common.Hash, repair bo
 		}
 	}
 
-	// Delete istanbul snapshot database further two epochs
-	var (
-		curBlkNum   = bc.CurrentBlock().Number().Uint64()
-		epoch       = bc.Config().Istanbul.Epoch
-		votingEpoch = curBlkNum - (curBlkNum % epoch)
-	)
-	if votingEpoch == 0 {
-		votingEpoch = 1
-	}
-	// Delete the snapshot state beyond the block number of the previous epoch on the right
-	for i := curBlkNum; i >= votingEpoch; i-- {
-		if params.IsCheckpointInterval(i) {
-			// delete from sethead number to previous two epoch block nums
-			// to handle a block that contains non-empty vote data to make sure
-			// the `HandleGovernanceVote()` cannot be skipped
-			bc.db.DeleteIstanbulSnapshot(bc.GetBlockByNumber(i).Hash())
+	if deleteSnpashot {
```

Comment:
How about putting this logic in the "else" branch below? Would it have the same effect?

https://github.com/klaytn/klaytn/blob/b62c83401a40934fade47f406993e23b792bb778/blockchain/blockchain.go#L635-L646

Reply:
This flag was to distinigush the callsite between them. If it's absent, every callsite of `SetHead` would delete snapshot data for further epoch. To remove this bad UX, I think the API call path only should trigger it.

Reply:
Initially, I thought `deleteSnapshot` is always the inverse of `repair`, because
- `SetHead` from API: `repair=false`, `deleteSnapshot=true`,
- otherwise: `repair=true`, `deleteSnapshot=false`.

In my understanding, `repair` flag indicates whether or not to delete block bodies and receipts. So, when we're deleting them, I thought snapshot may be deleted as well.

I didn't notice at the time of writing the first comment, but I've found some counterexamples where block bodies are deleted but snapshot is not (i.e., `repair=false` and `deleteSnapshot=false`):
- https://github.com/hyunsooda/klaytn/blob/sethead-snapshot-delete-flag/blockchain/blockchain.go#L330
- https://github.com/hyunsooda/klaytn/blob/sethead-snapshot-delete-flag/blockchain/blockchain.go#L790

I just wanted to make sure if such behavior is okay. Please correct me if I'm wrong. Thanks

Reply:
Indeed, the desired outcome of the snapshot deletion was initiated solely through the API call. As previously highlighted, this removes the time wasted on regenerating deleted snapshot data. Given that header-based voting is infrequent and `sethead` calls are frequent, this implementation avoids boring delays. Should the necessity for a snapshot deletion arise, the operator can promptly initiate the process by making a simple API call.

Reply:
I wanted to ask if we can reuse `repair` flag to achieve the goal, instead of introducing new `deleteSnapshot` flag. It seems to me the both flags serve the same purpose.

Reply:
No, it seems not identical. All three calls unconditionally remove snpashot because of `repair=false` in your suggestion.
- https://github.com/hyunsooda/klaytn/blob/sethead-snapshot-delete-flag/blockchain/blockchain.go#L330
- https://github.com/hyunsooda/klaytn/blob/sethead-snapshot-delete-flag/blockchain/blockchain.go#L790
- https://github.com/klaytn/klaytn/pull/2064/files#diff-de2bd0674bf71cd6e3d92cb5621577831308e4cc990a15933e5e853f3358e4e0R323

Reply:
Aren't they rare cases? It seems that first is when bad hash exists, second is when reset to genesis, and third is when genesis config is not compatible. In such error cases, I wanted to ask which is ok: to leave or not to leave the snapshots.

Reply:
Because block bodies are removed in such error cases, I thought it may be ok to delete snapshots as well. I may be wrong so I'm asking the question :)

Reply:
I realize your suggestion appears identical, so deleting everything in such cases seems reasonable. I also need to correct my comment: `sethead` is rare, but `repairing` is frequent (`repair=true`). The snapshot removal handler was triggered unconditionally, and I've addressed it with your suggestion. PTAL. Thank you.